rust_icu_uloc/src/lib.rs

```diff
@@ -63,55 +63,178 @@ impl TryFrom<&ffi::CStr> for ULoc {
+            buf.resize(full_len, 0);
+            String::from_utf8(buf).map_err(|_| common::Error::string_with_interior_nul())
+        }
+    }
+}
+
+buffered_string_method_with_retry!(
+    buffered_string_char_star,
+    LOCALE_CAPACITY,
+    [char_star: *const raw::c_char,],
+    []
+);
+
+buffered_string_method_with_retry!(
+    buffered_string_char_star_ubool,
+    LOCALE_CAPACITY,
+    [char_star: *const raw::c_char,],
+    [ubool: rust_icu_sys::UBool,]
+);
+
+macro_rules! call_buffered_string_char_star_with_repr {
+    ($self:ident, $uloc_method:ident) => {{
+        let asciiz = $self.as_c_string()?;
+        buffered_string_char_star(versioned_function!($uloc_method), asciiz.as_ptr())
+    }};
+}
+
 impl ULoc {
     /// Implements `uloc_getLanguage`.
     pub fn language(&self) -> Result<String, common::Error> {
-        self.call_buffered_string_method(versioned_function!(uloc_getLanguage))
+        call_buffered_string_char_star_with_repr!(self, uloc_getLanguage)
     }
 
     /// Implements `uloc_getScript`.
     pub fn script(&self) -> Result<String, common::Error> {
-        self.call_buffered_string_method(versioned_function!(uloc_getScript))
+        call_buffered_string_char_star_with_repr!(self, uloc_getScript)
     }
 
     /// Implements `uloc_getCountry`.
     pub fn country(&self) -> Result<String, common::Error> {
-        self.call_buffered_string_method(versioned_function!(uloc_getCountry))
+        call_buffered_string_char_star_with_repr!(self, uloc_getCountry)
     }
 
     /// Implements `uloc_getVariant`.
     pub fn variant(&self) -> Result<String, common::Error> {
-        self.call_buffered_string_method(versioned_function!(uloc_getVariant))
+        call_buffered_string_char_star_with_repr!(self, uloc_getVariant)
     }
 
     /// Implements `uloc_canonicalize` from ICU4C.
     pub fn canonicalize(&self) -> Result<ULoc, common::Error> {
-        self.call_buffered_string_method(versioned_function!(uloc_canonicalize))
-            .map(|repr| ULoc { repr })
+        call_buffered_string_char_star_with_repr!(self, uloc_canonicalize).map(|repr| ULoc { repr })
     }
 
     /// Implements `uloc_addLikelySubtags` from ICU4C.
     pub fn add_likely_subtags(&self) -> Result<ULoc, common::Error> {
-        self.call_buffered_string_method(versioned_function!(uloc_addLikelySubtags))
+        call_buffered_string_char_star_with_repr!(self, uloc_addLikelySubtags)
             .map(|repr| ULoc { repr })
     }
 
     /// Implements `uloc_minimizeSubtags` from ICU4C.
     pub fn minimize_subtags(&self) -> Result<ULoc, common::Error> {
-        self.call_buffered_string_method(versioned_function!(uloc_minimizeSubtags))
+        call_buffered_string_char_star_with_repr!(self, uloc_minimizeSubtags)
             .map(|repr| ULoc { repr })
     }
 
+    // Implements `uloc_toLanguageTag` from ICU4C.
+    pub fn to_language_tag(&self, strict: bool) -> Result<String, common::Error> {
+        let repr = self.as_c_string()?;
```

Comment:
Thinking aloud now.  When `as_c_string` fails,  is this a programmer error, or a user error?

I ask because it seems a bit weird that to_language_tag would need to throw an error.  This is an immutable function on `&self` which seems to indicate that we should be able to know in advance if `repr` contains a valid C string or not.  

In a way, it looks like stepping onto an error here means that our `ULoc` is already an invalid struct.  Which seems like something we should prevent.  WDYT?

Reply:
You're right -- we don't expose any way to create an invalid `ULoc`. Reverting the `as_c_str` change.